### PR TITLE
[SPARK-42353][SS] Cleanup orphan sst and log files in RocksDB checkpoint directory

### PR DESCRIPTION
**Pull request by chaoqin-li1123**
### What changes were proposed in this pull request?
When RocksDB version.zip file get overwritten (e.g. concurrent task execution, task/stage/batch reattempts) or the zip file don't get uploaded successfully, the associated sst and log files don't get garbage collected. We can clean up these SST files during periodic state store maintenance. The major concern is that sst files for ongoing version also appear to be "orphan" because they are uploaded before zip file, we have to be careful not to delete them. To be conservative, we only delete orphan files older than all files tracked by current metadata when there are at least 2 versions.

### Why are the changes needed?
When RocksDB version.zip file get overwritten (e.g. concurrent task execution, task/stage/batch reattempts) or the zip file don't get uploaded successfully, the associated sst and log files don't get garbage collected.([https://github.com/databricks/runtime/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala](https://github.com/databricks/runtime/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala#L305-L309)) These files consume storage. 

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Add unit test.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala [hunk 1]
```diff
@@ -269,29 +306,46 @@ class RocksDBFileManager(
...
-        val dfsFile = dfsFilePath(file.dfsFileName)
+        val dfsFile = dfsFilePath(dfsFileName)
         fm.delete(dfsFile)
-        logDebug(s"Deleted file $file that was last used in version $maxUsedVersion")
+        if (maxUsedVersion == -1) {
+          logDebug(s"Deleted orphan file $dfsFileName")
+        } else {
+          logDebug(s"Deleted file $dfsFileName that was last used in version $maxUsedVersion")
+        }
+        logDebug(s"Deleted file $dfsFileName that was last used in version $maxUsedVersion")
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala:340]: nit: remove?

**Review by HeartSaVioR — COMMENTED** [review]
Only nit. Thanks! Shall we remove WIP in the PR title as it seems to be ready to be reviewed and merged?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala (hunk 1 above)
Comment (chaoqin-li1123) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBFileManager.scala]: Thanks, fixed.

**Comment by chaoqin-li1123** [comment]
@rangadi

**Review by HeartSaVioR — APPROVED** [review]
+1 pending build.

**Comment by rangadi** [comment]
+1.

**Comment by HeartSaVioR** [comment]
Thanks! Merging to master.

**Pull request closed without merging**